### PR TITLE
[spv-out]: ability to separate entry points

### LINKED ISSUE #1328: [spv-out] Multiple entry points causes failure with Adreno drivers
Can be reproduced with a shader as simple as:

```wgsl
[[stage(fragment)]]
fn fs_main_linear(in: VertexOutput) -> FragmentOutput {
    return in.v_Color;
}

[[stage(fragment)]]
fn fs_main_srgb(in: VertexOutput) -> FragmentOutput {
    return in.v_Color;
}
```

Vulkan failure when creating render pipeline, simply `ERROR_UNKNOWN`, no other information.

(Tested with Adreno 620 on Pixel 5a)

### PR DESCRIPTION
**Pull request by kvark**
Closes #1328 (allows us to work around...)
Comes with a test!

### CODE REVIEW

**Comment by encounter** [comment]
This appears to generate invalid SPIR-V when functions are used that reference uniforms/samplers from a specific stage:

```wgsl
[[block]]
struct Uniforms {
    u_Matrix: mat4x4<f32>;
};

struct VertexInput {
    [[location(0)]] a_Pos: vec2<f32>;
    [[location(1)]] a_UV: vec2<f32>;
    [[location(2)]] a_Color: vec4<f32>;
};

struct VertexOutput {
    [[location(0)]] v_UV: vec2<f32>;
    [[location(1)]] v_Color: vec4<f32>;
    [[builtin(position)]] v_Position: vec4<f32>;
};

[[group(0), binding(0)]]
var<uniform> uniforms: Uniforms;

fn get_mtx() -> mat4x4<f32> {
    return uniforms.u_Matrix;
}

[[stage(vertex)]]
fn vs_main(in: VertexInput) -> VertexOutput {
    var out: VertexOutput;
    out.v_UV = in.a_UV;
    out.v_Color = in.a_Color;
    out.v_Position = get_mtx() * vec4<f32>(in.a_Pos.xy, 0.0, 1.0);
    return out;
}

[[group(1), binding(0)]]
var u_Texture: texture_2d<f32>;
[[group(1), binding(1)]]
var u_Sampler: sampler;

fn srgb_to_linear(srgb: vec4<f32>) -> vec4<f32> {
    let color_srgb = srgb.rgb;
    let selector = ceil(color_srgb - 0.04045); // 0 if under value, 1 if over
    let under = color_srgb / 12.92;
    let over = pow((color_srgb + 0.055) / 1.055, vec3<f32>(2.4));
    let result = mix(under, over, selector);
    return vec4<f32>(result, srgb.a);
}

[[stage(fragment)]]
fn fs_main_linear(in: VertexOutput) -> [[location(0)]] vec4<f32> {
    let color = srgb_to_linear(in.v_Color);
    return color * textureSample(u_Texture, u_Sampler, in.v_UV);
}

[[stage(fragment)]]
fn fs_main_srgb(in: VertexOutput) -> [[location(0)]] vec4<f32> {
    let color = in.v_Color;
    return color * textureSample(u_Texture, u_Sampler, in.v_UV);
}
```

For the fragment shaders, spirv-val fails with:
```
error: line 84: Id is 0
```
within the get_mtx function.

**Review by jimblandy — APPROVED** [review]
Looks good to me, just a concern about test output filename generation.

**Comment by kvark** [comment]
@encounter your early testing saved us and our users quite a bit of trouble. Thank you, the hero we need!

File: src/back/spv/writer.rs [hunk 1]
```diff
@@ -1278,22 +1279,44 @@ impl Writer {
...
         }
 
         // all functions
         for (handle, ir_function) in ir_module.functions.iter() {
             let info = &mod_info[handle];
+            if let Some(index) = ep_index {
+                let ep_info = mod_info.get_entry_point(index);
+                for (vh, _) in ir_module.global_variables.iter() {
+                    // if this function uses globals in any way that the
+                    // entry point isn't, then we skip it.
```
[the remark this answers is not in the document]
Comment (kvark) [reply src/back/spv/writer.rs]: Oh snap, kudos to the author of this function!
Not sure about the comment. Ideally, the function name should reflect this already. Will add your suggestion.

File: tests/snapshots.rs [hunk 2]
```diff
@@ -198,13 +201,26 @@ fn write_output_spv(
...
+        for ep in module.entry_points.iter() {
+            let pipeline_options = spv::PipelineOptions {
+                entry_point: ep.name.clone(),
+                shader_stage: ep.stage,
+            };
+            let spv = spv::write_vec(module, info, &options, Some(&pipeline_options)).unwrap();
+            let dis = rspirv::dr::load_words(spv)
+                .expect("Produced invalid SPIR-V")
+                .disassemble();
+            let path = format!("spv/{}.{:?}.spvasm", file_name, ep.stage);
```
[the remark this answers is not in the document]
Comment (kvark) [reply tests/snapshots.rs:214]: Great point!

File: src/back/spv/writer.rs [hunk 3]
```diff
@@ -1332,11 +1355,25 @@ impl Writer {
...
         ir_module: &crate::Module,
         info: &ModuleInfo,
+        pipeline_options: Option<&PipelineOptions>,
         words: &mut Vec<Word>,
     ) -> Result<(), Error> {
         self.reset();
 
-        self.write_logical_layout(ir_module, info)?;
+        // Try to find the entry point and corresponding index
+        let ep_index = match pipeline_options {
```
[the remark this answers is not in the document]
Comment (kvark) [reply src/back/spv/writer.rs:1366]: We could, but then we'd have to do this swapping of `Option<Result>` to `Result<Option>`, which make the code more complicated than it needs to be here.

**Pull request merged**